WordPress/openverse
# Disallow crawlers from high cardinality collection pages

## Fixes



Related to ongoing issue with crawlers going through the collection endpoints

## Description


Collection pages for creator and tag have unlimited cardinality. Crawlers end up spending heaps of time quickly going through the tag and creator pages. Source collections on the other hand have bounded cardinality based on the number of sources we've defined for each media type, so that is safe. However, we do not want to disallow bots _at all_ from these pages, so we cannot add them to robots.txt. Instead, we use the meta tags (and header, thanks to `@nuxtjs/robots` for handling this transparently) to tell crawlers it's okay to visit the page, but not to follow the links. This makes sure that OpenGraph tags are always accessible on relevant pages, and that pages we want indexed can still be indexed without causing a flood of new requests due to crawling results on those pages.

I've also added a new Playwright snapshot test so we can easily audit changes to the robots.txt in diffs of PRs. The locales reflect those enabled for Playwright, which are a subset used for testing. The actual file uses the real locales, and you can see it by following the testing instructions below. This new test supplements the route specific tests.

Finally, we had unnecessarily duplicative robots rules. Robots.txt rules are prefix based, so there's no need to have `/search` and `/search/images`. `/search` already covers everything that starts with search.

https://developers.google.com/search/docs/crawling-indexing/robots/robots_txt?csw=1

## Testing Instructions

Review the new robots.txt by:

1. Checking out the new snapshot in the playwright tests in the diff of this PR
2. Run the Nuxt dev server locally (`ov nuxt`) and visit <http://127.0.0.1:8443/robots.txt?mockProductionEnv>; this will show you the full robots.txt with all locales (if you have them downloaded, which you should if using `ov nuxt`).

Review the updated Playwright tests that prove the meta tags are appropriate for each of the relevant scenarios. You can also test it locally by visiting the pages and adding `?mockProductionEnv` to them. Otherwise you will get the default for local environments on pages that do not explicitly use `useRobotsRule` which is noindex,nofollow (import for checking the static pages, for example).

## Checklist


- [N/A] I ran the DAG documentation generator (`ov just catalog/generate-docs` for catalog
      PRs) or the media properties generator (`ov just catalog/generate-docs media-props`
      for the catalog or `ov just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
